File: clients/erlang/src/network_mutex_protocol.erl
-module(network_mutex_protocol).

-export([
    protocol_version/0,
    max_composite_keys/0,
    request_types/0,
    response_types/0,
    request_type_to_wire/1,
    response_type_from_wire/1,
    version_request/1,
    version_request/2,
    auth_request/2,
    lock_request_single/5,
    lock_request_composite/4,
    unlock_request_single/4,
    unlock_request_composite/3,
    rw_request/3,
    lock_info_request/2,
    ls_request/1,
    heartbeat_request/1
]).

protocol_version() -> "0.1.0".
max_composite_keys() -> 5.

request_types() ->
    [
        "version",
        "auth",
        "lock",
        "unlock",
        "registerRead",
        "registerWrite",
        "endRead",
        "endWrite",
        "lockInfo",
        "ls",
        "heartbeat"
    ].

response_types() ->
    [
        "version",
        "auth",
        "lock",
        "compositeLock",
        "unlock",
        "registerReadResult",
        "registerWriteResult",
        "endReadResult",
        "endWriteResult",
        "lockInfo",
        "lsResult",
        "reelection",
        "error",
        "ok"
    ].

request_type_to_wire(version) -> "version";
request_type_to_wire(auth) -> "auth";
request_type_to_wire(lock) -> "lock";
request_type_to_wire(unlock) -> "unlock";
request_type_to_wire(register_read) -> "registerRead";
request_type_to_wire(register_write) -> "registerWrite";
request_type_to_wire(end_read) -> "endRead";
request_type_to_wire(end_write) -> "endWrite";
request_type_to_wire(lock_info) -> "lockInfo";
request_type_to_wire(ls) -> "ls";
request_type_to_wire(heartbeat) -> "heartbeat".

response_type_from_wire("version") -> version;
response_type_from_wire("auth") -> auth;
response_type_from_wire("lock") -> lock;
response_type_from_wire("compositeLock") -> composite_lock;
response_type_from_wire("unlock") -> unlock;
response_type_from_wire("registerReadResult") -> register_read_result;
response_type_from_wire("registerWriteResult") -> register_write_result;
response_type_from_wire("endReadResult") -> end_read_result;
response_type_from_wire("endWriteResult") -> end_write_result;
response_type_from_wire("lockInfo") -> lock_info;
response_type_from_wire("lsResult") -> ls_result;
response_type_from_wire("reelection") -> reelection;
response_type_from_wire("error") -> error;
response_type_from_wire("ok") -> ok;
response_type_from_wire(_) -> unknown.

version_request(Uuid) -> version_request(Uuid, protocol_version()).

version_request(Uuid, Value) ->
    frame([{string, "type", request_type_to_wire(version)}, {string, "uuid", Uuid}, {string, "value", Value}]).

auth_request(Uuid, Token) ->
    frame([{string, "type", request_type_to_wire(auth)}, {string, "uuid", Uuid}, {string, "token", Token}]).

lock_request_single(Uuid, Key, TtlMs, MaxHolders, Wait) ->
    frame(
        compact([
            {string, "type", request_type_to_wire(lock)},
            {string, "uuid", Uuid},
            {string, "key", Key},
            optional_int("ttl", TtlMs, TtlMs > 0),
            optional_int("max", MaxHolders, is_integer(MaxHolders)),
            optional_bool("wait", Wait, is_boolean(Wait))
        ])
    ).

lock_request_composite(Uuid, Keys, TtlMs, Wait) ->
    Count = length(Keys),
    case Count >= 1 andalso Count =< ?MODULE:max_composite_keys() of
        true ->
            frame(
                compact([
                    {string, "type", request_type_to_wire(lock)},
                    {string, "uuid", Uuid},
                    {strings, "keys", Keys},
                    optional_int("ttl", TtlMs, TtlMs > 0),
                    optional_bool("wait", Wait, is_boolean(Wait))
                ])
            );
        false ->
            error({bad_composite_key_count, Count})
    end.

unlock_request_single(Uuid, Key, LockUuid, Force) ->
    frame(
        compact([
            {string, "type", request_type_to_wire(unlock)},
            {string, "uuid", Uuid},
            {string, "key", Key},
            optional_string("lockUuid", LockUuid, LockUuid =/= undefined andalso LockUuid =/= ""),
            optional_bool("force", Force, Force =:= true)
        ])
    ).

unlock_request_composite(Uuid, Keys, LockUuid) ->
    frame(
        compact([
            {string, "type", request_type_to_wire(unlock)},
            {string, "uuid", Uuid},
            {strings, "keys", Keys},
            optional_string("lockUuid", LockUuid, LockUuid =/= undefined andalso LockUuid =/= "")
        ])
    ).

rw_request(Type, Uuid, Key) ->
    frame([{string, "type", request_type_to_wire(Type)}, {string, "uuid", Uuid}, {string, "key", Key}]).

lock_info_request(Uuid, Key) -> rw_request(lock_info, Uuid, Key).

ls_request(Uuid) ->
    frame([{string, "type", request_type_to_wire(ls)}, {string, "uuid", Uuid}]).

heartbeat_request(Uuid) ->
    frame([{string, "type", request_type_to_wire(heartbeat)}, {string, "uuid", Uuid}]).

optional_string(Key, Value, true) -> {string, Key, Value};
optional_string(_, _, false) -> skip.

optional_int(Key, Value, true) -> {int, Key, Value};
optional_int(_, _, false) -> skip.

optional_bool(Key, Value, true) -> {bool, Key, Value};
optional_bool(_, _, false) -> skip.

compact(Fields) -> [Field || Field <- Fields, Field =/= skip].

frame(Fields) ->
    "{" ++ join([field_to_json(Field) || Field <- Fields], ",") ++ "}\n".

field_to_json({string, Key, Value}) ->
    quote(Key) ++ ":" ++ quote(Value);
field_to_json({int, Key, Value}) ->
    quote(Key) ++ ":" ++ integer_to_list(Value);
field_to_json({bool, Key, true}) ->
    quote(Key) ++ ":true";
field_to_json({bool, Key, false}) ->
    quote(Key) ++ ":false";
field_to_json({strings, Key, Values}) ->
    quote(Key) ++ ":[" ++ join([quote(Value) || Value <- Values], ",") ++ "]".

quote(Value) when is_binary(Value) -> quote(binary_to_list(Value));
quote(Value) ->
    "\"" ++ escape(Value) ++ "\"".

escape([]) -> [];
escape([$" | Rest]) -> [$\\, $" | escape(Rest)];
escape([$\\ | Rest]) -> [$\\, $\\ | escape(Rest)];
escape([$\n | Rest]) -> [$\\, $n | escape(Rest)];
escape([Char | Rest]) -> [Char | escape(Rest)].

join([], _) -> "";
join([One], _) -> One;
join([One | Rest], Sep) -> One ++ Sep ++ join(Rest, Sep).

File: clients/ocaml/network_mutex_protocol.ml
let protocol_version = "0.1.0"
let max_composite_keys = 5
let max_fencing_token = 9_007_199_254_740_991

type request_type =
  | Version
  | Auth
  | Lock
  | Unlock
  | RegisterRead
  | RegisterWrite
  | EndRead
  | EndWrite
  | LockInfo
  | Ls
  | Heartbeat

type response_type =
  | VersionResponse
  | AuthResponse
  | LockResponse
  | CompositeLockResponse
  | UnlockResponse
  | RegisterReadResult
  | RegisterWriteResult
  | EndReadResult
  | EndWriteResult
  | LockInfoResponse
  | LsResult
  | Reelection
  | ErrorResponse
  | OkResponse
  | UnknownResponse

let request_types =
  [ "version"; "auth"; "lock"; "unlock"; "registerRead"; "registerWrite";
    "endRead"; "endWrite"; "lockInfo"; "ls"; "heartbeat" ]

let response_types =
  [ "version"; "auth"; "lock"; "compositeLock"; "unlock";
    "registerReadResult"; "registerWriteResult"; "endReadResult";
    "endWriteResult"; "lockInfo"; "lsResult"; "reelection"; "error"; "ok" ]

let request_type_to_wire = function
  | Version -> "version"
  | Auth -> "auth"
  | Lock -> "lock"
  | Unlock -> "unlock"
  | RegisterRead -> "registerRead"
  | RegisterWrite -> "registerWrite"
  | EndRead -> "endRead"
  | EndWrite -> "endWrite"
  | LockInfo -> "lockInfo"
  | Ls -> "ls"
  | Heartbeat -> "heartbeat"

let response_type_from_wire = function
  | "version" -> VersionResponse
  | "auth" -> AuthResponse
  | "lock" -> LockResponse
  | "compositeLock" -> CompositeLockResponse
  | "unlock" -> UnlockResponse
  | "registerReadResult" -> RegisterReadResult
  | "registerWriteResult" -> RegisterWriteResult
  | "endReadResult" -> EndReadResult
  | "endWriteResult" -> EndWriteResult
  | "lockInfo" -> LockInfoResponse
  | "lsResult" -> LsResult
  | "reelection" -> Reelection
  | "error" -> ErrorResponse
  | "ok" -> OkResponse
  | _ -> UnknownResponse

let validate_fencing_token token =
  if token < 1 || token > max_fencing_token then
    Error "fencingToken must be an exact integer in the authority domain"
  else Ok token

let fencing_token_from_response fields =
  match List.assoc_opt "fencingToken" fields with
  | None -> Error "missing fencingToken"
  | Some token -> validate_fencing_token token

let fencing_tokens_from_response keys fields =
  if keys = [] then Error "composite grant requires at least one key"
  else if List.length (List.sort_uniq String.compare keys) <> List.length keys then
    Error "composite grant keys must be unique"
  else if List.length fields <> List.length keys then
    Error "fencingTokens/key cardinality mismatch"
  else
    let rec loop acc = function
      | [] -> Ok (List.rev acc)
      | key :: rest ->
          (match List.assoc_opt key fields with
           | None -> Error ("missing fencingToken for key " ^ key)
           | Some token ->
               match validate_fencing_token token with
               | Error why -> Error (why ^ ": " ^ key)
               | Ok exact -> loop ((key, exact) :: acc) rest)
    in
    loop [] keys

let json_escape s =
  let b = Buffer.create (String.length s) in
  String.iter
    (function
      | '"' -> Buffer.add_string b "\\\""
      | '\\' -> Buffer.add_string b "\\\\"
      | '\n' -> Buffer.add_string b "\\n"
      | c -> Buffer.add_char b c)
    s;
  Buffer.contents b

let quote s = "\"" ^ json_escape s ^ "\""
let field_string key value = quote key ^ ":" ^ quote value
let field_int key value = quote key ^ ":" ^ string_of_int value
let field_bool key value = quote key ^ ":" ^ if value then "true" else "false"
let field_strings key values = quote key ^ ":[" ^ String.concat "," (List.map quote values) ^ "]"
let frame fields = "{" ^ String.concat "," fields ^ "}\n"

let version_request ?(value = protocol_version) uuid =
  frame [ field_string "type" "version"; field_string "uuid" uuid; field_string "value" value ]

let auth_request uuid token =
  frame [ field_string "type" "auth"; field_string "uuid" uuid; field_string "token" token ]

let lock_request_single ?pid ?(ttl = 0) ?max ?wait uuid key =
  let fields = [ field_string "type" "lock"; field_string "uuid" uuid; field_string "key" key ] in
  let fields = match pid with Some value -> fields @ [ field_int "pid" value ] | None -> fields in
  let fields = if ttl > 0 then fields @ [ field_int "ttl" ttl ] else fields in
  let fields = match max with Some value -> fields @ [ field_int "max" value ] | None -> fields in
  let fields = match wait with Some value -> fields @ [ field_bool "wait" value ] | None -> fields in
  frame fields

let lock_request_composite ?(ttl = 0) ?wait uuid keys =
  let count = List.length keys in
  if count < 1 || count > max_composite_keys then invalid_arg "composite key count must be 1..=5";
  let fields = [ field_string "type" "lock"; field_string "uuid" uuid; field_strings "keys" keys ] in
  let fields = if ttl > 0 then fields @ [ field_int "ttl" ttl ] else fields in
  let fields = match wait with Some value -> fields @ [ field_bool "wait" value ] | None -> fields in
  frame fields

let unlock_request_single ?(force = false) uuid key lock_uuid =
  let fields = [ field_string "type" "unlock"; field_string "uuid" uuid; field_string "key" key ] in
  let fields = if lock_uuid <> "" then fields @ [ field_string "lockUuid" lock_uuid ] else fields in
  let fields = if force then fields @ [ field_bool "force" true ] else fields in
  frame fields

let unlock_request_composite uuid keys lock_uuid =
  let fields = [ field_string "type" "unlock"; field_string "uuid" uuid; field_strings "keys" keys ] in
  let fields = if lock_uuid <> "" then fields @ [ field_string "lockUuid" lock_uuid ] else fields in
  frame fields

let rw_request request_type uuid key =
  frame [ field_string "type" (request_type_to_wire request_type); field_string "uuid" uuid; field_string "key" key ]

let lock_info_request uuid key = rw_request LockInfo uuid key
let ls_request uuid = frame [ field_string "type" "ls"; field_string "uuid" uuid ]
let heartbeat_request uuid = frame [ field_string "type" "heartbeat"; field_string "uuid" uuid ]
